reflect/src/main/java/dagger/reflect/Scope.java

```diff
@@ -380,13 +382,49 @@ Scope build() {
         replaced =
             allBindings.put(mapOfProviderKey, new UnlinkedMapOfProviderBinding(entryBindings));
         if (replaced != null) {
-          throw new IllegalStateException(); // TODO implicit map binding duplicates explicit one.
+          throw new IllegalStateException(
+              buildDuplicateMapBindingMessage(mapOfProviderKey, entryBindings, replaced));
         }
       }
 
       return new Scope(allBindings, jitLookupFactory, annotations, parent);
     }
 
+    private static String buildDuplicateSetBindingMessage(
+        Key key, SetBindings setBindings, Binding replaced) {
+      StringBuilder builder = new StringBuilder("[Dagger/DuplicateBindings] ");
+      builder.append(key).append(" has incompatible bindings or declarations:\n");
+      builder.append("  Set bindings and declarations:\n");
+      for (Binding elementBinding : setBindings.elementBindings) {
+        builder.append("    @IntoSet ").append(elementBinding).append("\n");
```

Comment:
`Scope` shouldn't know anything about Dagger. Can we change these messages in a way to provide this information without encoding the implementation details of _how_ Dagger contributes bindings into a set? Should the binding itself report this information in its `toString()` perhaps?

Reply:
The `Binding` doesn't know it came from `@IntoSet/@ElementsIntoSet/@IntoMap`:
![image](https://user-images.githubusercontent.com/2906988/62499188-a7a67100-b7d9-11e9-88d6-a61019020c25.png)
unless we look into `method.declaredAnnotations`.

I removed `dagger.*` class references and the dagger error code and fixed sentence comments in [force push](https://github.com/JakeWharton/dagger-reflect/compare/c89290e521e6542d6ae13fcf714c857936bbcb83..db6b03c957bc3b9bcbf99feb7c75e9baf2891c6b)

Reply:
Looking at Dagger output that's what's happening there too, happy to open a followup changing the toString of `/(un)?linked(provides|binds)binding/` to include the full signature including annotations and parameters.

---

Did a quick test with:
```diff
  public String toString() {
    return "@Provides["
+        + Arrays.toString(method.getDeclaredAnnotations())
        + method.getDeclaringClass().getName() + '.' + method.getName() + "(…)]";
  }
```
```
  Set bindings and declarations:
    [single  ] @Binds[[@dagger.Binds(), @dagger.multibindings.IntoSet()]com.example.MultibindingSetConflict$Module1.three(â€¦)]
    [single  ] @Provides[[@dagger.Provides(), @dagger.multibindings.IntoSet()]com.example.MultibindingSetConflict$Module1.one(â€¦)]
    [multiple] @Provides[[@dagger.Provides(), @dagger.multibindings.ElementsIntoSet()]com.example.MultibindingSetConflict$Module1.two(â€¦)]
    [multiple] @Binds[[@dagger.Binds(), @dagger.multibindings.ElementsIntoSet()]com.example.MultibindingSetConflict$Module1.four(â€¦)]
```
I think I would need to remove the Binding's own annotation, and the noisy bits like `()` and known package names like dagger/dagger.mulitbindings/javax.inject to make it useful.

Reply:
You don't need to pull it from the method. We can supply an enum value or string at the time of binding creation which helps format this message.